plugin/trino-functions-python/src/main/java/io/trino/plugin/functions/python/TrinoTypes.java

```diff
@@ -0,0 +1,516 @@
+                    output.writeLong(picosToMicros(time.getPicoseconds()));
+                    output.writeShort(time.getOffsetMinutes());
+                }
+            }
+            case TimestampType timestampType -> output.writeLong(timestampType.isShort()
+                    ? (long) value
+                    : timestampToMicros((LongTimestamp) value));
+            case TimestampWithTimeZoneType timestampType -> {
+                if (timestampType.isShort()) {
+                    long packed = (long) value;
+                    long millis = unpackMillisUtc(packed);
+                    output.writeLong(millis * MICROSECONDS_PER_MILLISECOND);
+                    output.writeShort(zoneOffsetMinutes(millis, unpackZoneKey(packed).getKey()));
+                }
+                else {
+                    LongTimestampWithTimeZone timestamp = (LongTimestampWithTimeZone) value;
+                    long micros = timestamp.getEpochMillis() * MICROSECONDS_PER_MILLISECOND;
+                    output.writeLong(micros + picosToMicros(timestamp.getPicosOfMilli()));
+                    output.writeShort(zoneOffsetMinutes(timestamp.getEpochMillis(), timestamp.getTimeZoneKey()));
+                }
+            }
+            default -> javaToBinarySimple(type, value, output);
+        }
+    }
+
+    @SuppressWarnings("DuplicateBranchesInSwitch")
+    private static void javaToBinarySimple(Type type, Object value, SliceOutput output)
+    {
+        switch (type.getBaseName()) {
+            case StandardTypes.BOOLEAN -> output.writeByte((boolean) value ? 1 : 0);
+            case StandardTypes.BIGINT -> output.writeLong((long) value);
+            case StandardTypes.INTEGER -> output.writeInt(toIntExact((long) value));
+            case StandardTypes.SMALLINT -> output.writeShort(toIntExact((long) value));
+            case StandardTypes.TINYINT -> output.writeByte(toIntExact((long) value));
+            case StandardTypes.DOUBLE -> output.writeDouble((double) value);
+            case StandardTypes.REAL -> output.writeInt(toIntExact((long) value));
+            case StandardTypes.DATE -> output.writeInt(toIntExact((long) value));
+            case StandardTypes.TIME -> output.writeLong(picosToMicros((long) value));
+            case StandardTypes.INTERVAL_YEAR_TO_MONTH -> output.writeInt(toIntExact((long) value));
+            case StandardTypes.INTERVAL_DAY_TO_SECOND -> output.writeLong((long) value);
+            case StandardTypes.UUID,
+                 StandardTypes.IPADDRESS -> output.writeBytes((Slice) value);
+            case StandardTypes.VARCHAR,
+                 StandardTypes.VARBINARY,
+                 StandardTypes.JSON -> writeVariableSlice((Slice) value, output);
+            default -> throw new TrinoException(NOT_SUPPORTED, "Unsupported type: " + type);
+        }
+    }
+
+    private static void blockToBinary(Type type, Block block, int position, SliceOutput output)
```

Comment:
I would make this take ValueBlock, and then unwrap the value block in the caller code.  I can help you with this.

Reply:
We can follow up on this.